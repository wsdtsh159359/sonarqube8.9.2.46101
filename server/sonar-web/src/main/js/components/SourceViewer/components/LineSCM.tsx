/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Dropdown from 'sonar-ui-common/components/controls/Dropdown';
import { PopupPlacement } from 'sonar-ui-common/components/ui/popups';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import SCMPopup from './SCMPopup';

export interface LineSCMProps {
  line: T.SourceLine;
  previousLine: T.SourceLine | undefined;
}

export function LineSCM({ line, previousLine }: LineSCMProps) {
  const hasPopup = !!line.line;
  const cell = (
    <div className="source-line-scm-inner">
      {isSCMChanged(line, previousLine) ? line.scmAuthor || 'â€¦' : ' '}
    </div>
  );

  if (hasPopup) {
    let ariaLabel = translate('source_viewer.click_for_scm_info');
    if (line.scmAuthor) {
      ariaLabel = `${translateWithParameters(
        'source_viewer.author_X',
        line.scmAuthor
      )}, ${ariaLabel}`;
    }

    return (
      <td className="source-meta source-line-scm" data-line-number={line.line}>
        <Dropdown overlay={<SCMPopup line={line} />} overlayPlacement={PopupPlacement.RightTop}>
          <div aria-label={ariaLabel} role="button">
            {cell}
          </div>
        </Dropdown>
      </td>
    );
  } else {
    return (
      <td className="source-meta source-line-scm" data-line-number={line.line}>
        {cell}
      </td>
    );
  }
}

function isSCMChanged(s: T.SourceLine, p: T.SourceLine | undefined) {
  let changed = true;
  if (p != null && s.scmRevision != null && p.scmRevision != null) {
    changed = s.scmRevision !== p.scmRevision || s.scmDate !== p.scmDate;
  }
  return changed;
}

export default React.memo(LineSCM);
