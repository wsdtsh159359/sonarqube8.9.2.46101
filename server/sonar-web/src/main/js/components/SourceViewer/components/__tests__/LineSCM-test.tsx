/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { LineSCM, LineSCMProps } from '../LineSCM';

it('should render correctly', () => {
  const scmInfo = { scmRevision: 'foo', scmAuthor: 'foo', scmDate: '2017-01-01' };

  expect(shallowRender()).toMatchSnapshot('default');
  expect(
    shallowRender({ line: { line: 3, ...scmInfo }, previousLine: { line: 2, ...scmInfo } })
  ).toMatchSnapshot('same commit');
  expect(shallowRender({ line: { line: 3, scmDate: '2017-01-01' } })).toMatchSnapshot('no author');
});

function shallowRender(props: Partial<LineSCMProps> = {}) {
  return shallow(
    <LineSCM
      line={{ line: 3, scmAuthor: 'foo', scmDate: '2017-01-01' }}
      previousLine={{ line: 2, scmAuthor: 'bar', scmDate: '2017-01-02' }}
      {...props}
    />
  );
}
