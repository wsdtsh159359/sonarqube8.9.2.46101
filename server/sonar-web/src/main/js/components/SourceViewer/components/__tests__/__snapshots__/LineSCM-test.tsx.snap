// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<td
  className="source-meta source-line-scm"
  data-line-number={3}
>
  <Dropdown
    overlay={
      <Memo(SCMPopup)
        line={
          Object {
            "line": 3,
            "scmAuthor": "foo",
            "scmDate": "2017-01-01",
          }
        }
      />
    }
    overlayPlacement="right-top"
  >
    <div
      aria-label="source_viewer.author_X.foo, source_viewer.click_for_scm_info"
      role="button"
    >
      <div
        className="source-line-scm-inner"
      >
        foo
      </div>
    </div>
  </Dropdown>
</td>
`;

exports[`should render correctly: no author 1`] = `
<td
  className="source-meta source-line-scm"
  data-line-number={3}
>
  <Dropdown
    overlay={
      <Memo(SCMPopup)
        line={
          Object {
            "line": 3,
            "scmDate": "2017-01-01",
          }
        }
      />
    }
    overlayPlacement="right-top"
  >
    <div
      aria-label="source_viewer.click_for_scm_info"
      role="button"
    >
      <div
        className="source-line-scm-inner"
      >
        â€¦
      </div>
    </div>
  </Dropdown>
</td>
`;

exports[`should render correctly: same commit 1`] = `
<td
  className="source-meta source-line-scm"
  data-line-number={3}
>
  <Dropdown
    overlay={
      <Memo(SCMPopup)
        line={
          Object {
            "line": 3,
            "scmAuthor": "foo",
            "scmDate": "2017-01-01",
            "scmRevision": "foo",
          }
        }
      />
    }
    overlayPlacement="right-top"
  >
    <div
      aria-label="source_viewer.author_X.foo, source_viewer.click_for_scm_info"
      role="button"
    >
      <div
        className="source-line-scm-inner"
      >
         
      </div>
    </div>
  </Dropdown>
</td>
`;
